cbor.lua

```diff
@@ -127,6 +127,46 @@ end
 
 _VERSION = cbor_c._VERSION
 
+-- **********************************************************************
+-- usage: is_array = table_is_array(table_value)
+-- desc:        Check if table should be encoded as MAP or as ARRAY
+-- input:       table_value (table) table to analyze
+-- return:      is_array (bool) is the table looks like ARRAY or MAP
+-- ***********************************************************************
+
+local function table_is_array(t)
+    local is_array = true
+    local is_empty = true
+    for i,_ in pairs(t) do
+        is_empty = false
+        if type(i) ~= "number" or i <= 0 or i % 1 ~= 0 then
+            is_array = false
+            break
+        end
+    end
+    if is_empty then
+        is_array = false
+    end
+    return is_array
+end
```

Comment:
For me, this functions fails when t is a sparse array, like 

    { [1] = 1 , [10] = 10 , [100] = 100 }

There is a way to encode data specifically, and for the case of a sparse array, you can add the `__tocbor` metamethod to force a particular encoding for a table.

    local a = { [1] = 1 , [10] = 10 , [100] = 100 , n = 100 }

    setmetatable(a,{ __tocbor = function(array,sref,stref)
      local blob = cbor.TYPE.ARRAY(array.n)
      for i = 1 , array.n do
        blob = blob .. cbor.encode(array[i])
      end
      return blob
    end
    })

The `__tocbor` metamethod is supported in both org.conman.cbor and org.conman.cbor_s.  While it *is* mentioned in the `README` I'll admit it's brief and could use some expansion.  Also, for future reference, Lua 5.3 has the math.type() function, so line 142 could read:

    if math.type(i) ~= 'integer' or i <= 0 then

I don't think I want to pull this code in if there is a way it can be done without modifying the code.

Reply:
I was sure you'll accept this request and probably reject the other one :)

I'm currently using lua 5.1.4, it works as planned and compatible with your old code for your example (you use `#x` to check if the table is array). 
```
> x={ [1]=1, [10]=10, [100]=100 }
> print(table_is_array(x), #x)
true    1
```
The different is in mixed cases, it is more correct that old code:
```
> x={ [1]=1, [10] = 10, x = "X", [100] = 100}
> print(table_is_array(x), #x)
false   1
```

```
> x={ [1]=nil, [10]=10, [100]=100 }
> print(table_is_array(x), #x)
true    0
```
So, now these strange (in lua) cases are encoded in a strange way, with loosing data:
```
> print(hex(cbor.encode({ [1]=1, [10]=10, [100]=100 })))
81 01   2
> print(hex(cbor.encode({ [1]=nil, [10]=10, [100]=100 })))
a2 18 64 18 64 0a 0a    7
> return hex(cbor.encode({ [1]=1, [10]=10, x="X", [100]=100 }))
81 01   2
```
The second case is correct, but it uses MAP instead of ARRAY. It is not bug, only developer knows the real type (MAP or ARRAY) of this table. The other cases loose data and are wrong anyway.

I know about `__to_cbor` method. My first goal is to fix wrong encoding for sparse and mixed arrays. The second goal is to detect arrays more precisely then `#x` operation does because you can't trust it's results for sparce arrays or some key-value maps with `[1]` key. You don't need to tune most tables with `__to_cbor` when the second goal is complete. Also you can encode most of the data you know nothing about (for example, in JSON->CBOR converter).

So, your opinion is still to reject?

Reply:
I did agonize over sparse and mixed arrays when writing the code, but decided to not bother with trying to detect them, as it was a very hard problem and easy to get wrong for some set of people.  That's why I added the `__tocbor` method, as a way for the programmer to deal with difficult to encode Lua tables.  

The other issue is your method of finding the highest integer key in a Lua table.  In the worse case, it's O(n) if given an proper Lua array without holes, which means the less sparse the array, the longer it takes to find the length (Lua itself uses a binary search to keep it O(lgn)).  I do try some heuristics to determine when to use an ARRAY or MAP, but that's the issue---it's a heuristic.  

I'm more inclined to just document the issue and leave it at that---it's been four years since I wrote this module, and this is the first time this particular issue has come up.  I don't know how many people are actually using this module (I know I am at work) so it may be more of an issue that I realize, but it may not be worth solving it in the general case when there are workarounds.

Reply:
> O(n)
Yes, it's a huge problem and I don't like it. When I tried to do something with the problem, first I tried to use table.maxn. I know it is deprecated so I looked at the source code to rewrite it. The idea in the function is the same. Seems it is a reason they deprecated it. I decided that if such algorythm was implemented in lua core, I can do the same. 
Okey, I understand your point about this. For now 1.3.0 is good enough to solve my task, thanks!

Reply:
You're welcome.  One more thing---if you know you are working with a sparse array, you could always add a metatable with the `__len` and `__newindex` methods to track the highest integer key and avoid a lengthy algorithm to find it.  Then you won't even need a `__tocbor` method.  

But I'm glad you can work around this issue.